dilated-file-transfer.md

```diff
@@ -0,0 +1,499 @@
+
+Note that `"send"` and `"receive"` modes will still use Dilation as all clients supporting this protocol must.
+If a peer sends no version information at all, it will be using the classic protocol (and is thus using Transit and not Dilation for the peer-to-peer connection).
+
+The `"features"` key points at a list of message-formats / features understood by the peer.
+This allows for existing messages to be extended, or for new message types to be added.
+Peers MUST _accept_ messages for any features they declare in `"features"`.
+Peers MUST only send messages / attributes for features in the other side's list.
+Since there are only the core features currently, the only valid value is an empty list.
+Peers MUST expect any strings in this list in the future (e.g. if a new feature is added, the protocol version isn't necessarily bumped).
+
+   XXX:: maybe just lean on "version" for now? e.g. version `2` could introduce "features"?
+
+See "Example of Protocol Expansion" below for discussion about adding new attributes (including when we might increment the `"version"` instead of adding a new `"feature"`).
+
+
+## Protocol Details
+
+See the Dilation document for details on the Dilation setup procedure.
+Once a Dilation-supporting connection is open, we will have a "control" subchannel (subchannel #0).
+Either peer can also open additional subchannels.
+
+All control-channel messages are encoded using `msgpack`.
+   --> XXX: see "message encoding" discussion
+
+Control-channel message formats are described using Python pseudo-code to illustrate the data types involved.
+They are actually an encoded `Map` with `String` keys (to use `msgpack` lingo) and values as per the pseudo-code.
+
+All control-channel messages contain a integer "kind" field describing the sort of message it is.
+(That is, `"kind": "text"` for example, not the single-byte tag used for subchannel messages)
+
+Rejected idea: Version message, because we already do version negotiation via mailbox features.
+
+Rejected idea: Offer/Answer messages via the control channel: we need to open a subchannel anyway and the subchannel-IDs are not intended to be part of the Dilation public API.
+
+
+### Control Channel Messages
+
+Each side MAY send a free-form text message at any time.
+These messages look like:
+
+```python
+class Message:
+    message: str     # unicode string
+    kind: str = "text"
+```
+
+
+### Making an Offer
+
+Either side MAY propose any number of Offers at any time after the connection is set up.
+If the other peer specified `"mode": "send"` then this peer MUST NOT make any Offers.
+If this peer specified `"mode": "receive"` then this peer MUST NOT make any Offers.
+
+To make an Offer the peer opens a subchannel.
```

Comment:
There should be more clarification here what a subchannel is and how to open it. The current dilation spec is very vague about it and uses many python-isms I don't understand. If this is not the right place for an explanation then it should link to an explanation that explains it in a clear and portable way.

Reply:
I think anything that's unclear in the Dilation spec should indeed be fixed (but also that's really a separate discussion). Can you open a ticket about what you think it vague? (I've recently re-read https://github.com/magic-wormhole/magic-wormhole/blob/master/docs/dilation-protocol.md several times but also know Python ;) and the implementation so hard to know what's confusing now).

In short, it's a multiplexed channel down the peer-to-peer connection.

Reply:
The currently written version is more *documentation* of the Python implementation than a standalone *specification*. I have on my task list to rework the file once I know more about the implementation, although this won't help you better understand things right now.

Reply:
Spec seems clear here to me? L3 (layer 3) sends a `0x02 OPEN, 4-byte subchannel-id, 4-byte seqnum` message to open a subchannel and the "L4 Protocol" section explains what happens on these. Anyway, Dilation should be clarified etc elsewhere...